XenAdmin/Wizards/PatchingWizard/PlanActions/UploadPatchToMasterPlanAction.cs

```diff
@@ -31,99 +31,234 @@
+using System.IO;
 using XenAdmin.Actions;
 using XenAdmin.Core;
 using XenAPI;
 using System.Linq;
+using System.Windows.Forms;
+using XenAdmin.Dialogs;
 using XenAdmin.Network;
 
 
 namespace XenAdmin.Wizards.PatchingWizard.PlanActions
 {
     class UploadPatchToMasterPlanAction : PlanActionWithSession
     {
-        private readonly XenServerPatch patch;
-        private readonly List<PoolPatchMapping> mappings;
+        private readonly XenServerPatch xenServerPatch;
+        private readonly List<HostUpdateMapping> mappings;
         private Dictionary<XenServerPatch, string> AllDownloadedPatches = new Dictionary<XenServerPatch, string>();
         private KeyValuePair<XenServerPatch, string> patchFromDisk;
         private AsyncAction inProgressAction;
+        private bool skipDiskSpaceCheck;
+        private string updateFilePath;
+        private readonly Control invokingControl;
+        private readonly List<Host> selectedServers;
 
-        public UploadPatchToMasterPlanAction(IXenConnection connection, XenServerPatch patch, List<PoolPatchMapping> mappings, Dictionary<XenServerPatch, string> allDownloadedPatches, KeyValuePair<XenServerPatch, string> patchFromDisk)
+        public UploadPatchToMasterPlanAction(Control invokingControl, IXenConnection connection, XenServerPatch xenServerPatch,
+            List<HostUpdateMapping> mappings, Dictionary<XenServerPatch, string> allDownloadedPatches,
+            KeyValuePair<XenServerPatch, string> patchFromDisk, bool skipDiskSpaceCheck = false)
             : base(connection)
         {
-            this.patch = patch;
+            this.invokingControl = invokingControl;
+            this.xenServerPatch = xenServerPatch;
             this.mappings = mappings;
             AllDownloadedPatches = allDownloadedPatches;
             this.patchFromDisk = patchFromDisk;
+            this.skipDiskSpaceCheck = skipDiskSpaceCheck;
         }
 
-        protected override void RunWithSession(ref Session session)
+        public UploadPatchToMasterPlanAction(Control invokingControl, IXenConnection connection, List<Host> selectedServers,
+            string updateFilePath, List<HostUpdateMapping> mappings, bool skipDiskSpaceCheck = false)
+            : base(connection)
         {
-            var path = AllDownloadedPatches.ContainsKey(patch)
-                ? AllDownloadedPatches[patch]
-                : patchFromDisk.Key == patch ? patchFromDisk.Value : null;
-
-            var poolPatches = new List<Pool_patch>(session.Connection.Cache.Pool_patches);
-            var poolUpdates = new List<Pool_update>(session.Connection.Cache.Pool_updates);
+            this.invokingControl = invokingControl;
+            this.updateFilePath = updateFilePath;
+            this.mappings = mappings;
+            this.skipDiskSpaceCheck = skipDiskSpaceCheck;
+            this.selectedServers = selectedServers;
+        }
 
+        protected override void RunWithSession(ref Session session)
+        {

```

Comment:
This function is now quite long and it is hard hard to follow the code with all the nested `if`s. I wonder if we can extract some of the code in functions, to aid readability. For example, the first section, where we determine if there is an existing mapping, can be put in a separate function.